Library/Homebrew/service.rb

```diff
@@ -497,5 +500,93 @@ def to_systemd_timer
 
       timer + options.join("\n")
     end
+
+    # Only evaluate the service block once.
+    sig { void }
+    def eval_service_block
+      return if @eval_service_block
+
+      instance_eval(&@service_block)
+      @eval_service_block = true
+    end
+
+    # Recursively prepare the service hash for inclusion in the formula API JSON.
+    # - Replace each Symbol with a String prefixed by ":"
+    # - Replace the local HOMEBREW_PREFIX with "$HOMEBREW_PREFIX"
+    # - Replace the local home directory with "$HOME"
+    def serialize(elem = to_h)
+      case elem
+      when String, Pathname
+        elem.to_s.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
+            .gsub(Dir.home, "$HOME")
```

Comment:
Feels like having a shared method to do this in `Utils` or something that can be used elsewhere will help us avoid forgetting to add more of these.

Reply:
How would that method work? The way I see it we don't always apply or need to apply the same substitutions everywhere. We do have the `keg_relocate` code but something like that seems heavy handed for this specific case.

CC: @Rylan12 since you mentioned something similar in another thread

Reply:
We could also do the substitutions right before or immediately after writing the `cask.json` or `formula.json` file if we wanted to have something super generic and only do the substitutions in one place.

Reply:
I think one option is to have a method that just does all of the replacements we deem necessary (essentially `Cask::Cask.to_h_string_gsubs` before it was removed):

```ruby
def replace_strings(string)
  string.gsub(HOMEBREW_PREFIX, HOMEBREW_PREFIX_PLACEHOLDER)
        .gsub(HOME, HOME_PLACEHOLDER)
```

And then we can either selectively call this where it's needed or just recursively call this on the hash generated by `Formula#to_hash` before returning

Reply:
> just recursively call this on the hash generated by `Formula#to_hash` before returning

Something like this makes most sense to me.

I don't think it makes sense to scatter the replacements all over the codebase when they are only needed when reading/writing hashes.

CC @issyl0 as this affects https://github.com/Homebrew/brew/pull/15004 and possibly other PRs you have open

Reply:
> Yeah, the only reason I can see for not doing that is that there is at least one place we decided we didn't want it to be replaced in casks (#14595)

Yep, this is the exact reason I'm worried about applying it more generally. Also, if we know the substitutions are not needed for most of the formula hash we're just doing extra work for no reason.

Reply:
Yeh, I forgot this method was changed in https://github.com/Homebrew/brew/pull/14615 to use monkeypatches to avoid the need for as many explicit substitutions.

I think it'd be nicer to use this approach instead wherever possible.

Reply:
That could work for the hash generation but we'd still need to substitute things afterwards when deserializing. I think that the generating hash code for casks can probably just be converted into a generic mixin.

Reply:
Addressed in https://github.com/Homebrew/brew/pull/15007/commits/38146893c3966d0ad1a359f866df4dbf49b9e176